internal/ethapi/api.go

```diff
@@ -1299,6 +1299,90 @@ func newRPCTransactionFromBlockHash(b *types.Block, hash common.Hash) *RPCTransa
 	return nil
 }
 
+// accessListResult returns an optional accesslist
+// Its the result of the `debug_createAccessList` RPC call.
+// It contains an error if the transaction itself failed.
+type accessListResult struct {
+	Accesslist *types.AccessList `json:"accessList"`
+	Error      string            `json:"error,omitempty"`
+	GasUsed    hexutil.Uint64    `json:"gasUsed"`
+}
+
+// CreateAccessList creates a EIP-2930 type AccessList for the given transaction.
+// Reexec and BlockNrOrHash can be specified to create the accessList on top of a certain state.
+func (s *PublicBlockChainAPI) CreateAccessList(ctx context.Context, args SendTxArgs, blockNrOrHash *rpc.BlockNumberOrHash) (*accessListResult, error) {
+	bNrOrHash := rpc.BlockNumberOrHashWithNumber(rpc.PendingBlockNumber)
+	if blockNrOrHash != nil {
+		bNrOrHash = *blockNrOrHash
+	}
+	acl, gasUsed, vmerr, err := AccessList(ctx, s.b, bNrOrHash, args)
+	if err != nil {
+		return nil, err
+	}
+	result := &accessListResult{Accesslist: &acl, GasUsed: hexutil.Uint64(gasUsed)}
+	if vmerr != nil {
+		result.Error = vmerr.Error()
+	}
+	return result, nil
+}
+
+// AccessList creates an access list for the given transaction.
+// If the accesslist creation fails an error is returned.
+// If the transaction itself fails, an vmErr is returned.
+func AccessList(ctx context.Context, b Backend, blockNrOrHash rpc.BlockNumberOrHash, args SendTxArgs) (acl types.AccessList, gasUsed uint64, vmErr error, err error) {
+	db, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if db == nil || err != nil {
+		return nil, 0, nil, err
+	}
+	if err := args.setDefaults(ctx, b); err != nil {
+		// most likely an error in gas estimation, set gas to the block gas limit
+		log.Warn("Setting defaults failed", "error", err)
+		maxGas := hexutil.Uint64(header.GasLimit)
+		args.Gas = &maxGas
+	}
+	var (
+		gas        uint64
+		accessList types.AccessList
+		msg        types.Message
+	)
+	if args.AccessList != nil {
+		accessList = *args.AccessList
+	}
+	for i := 0; i < 10; i++ {
+		log.Trace("Creating access list", "accesslist", accessList)
+		// Copy the original db so we don't modify it
+		statedb := db.Copy()
+		msg = types.NewMessage(args.From, args.To, uint64(*args.Nonce), args.Value.ToInt(), uint64(*args.Gas), args.GasPrice.ToInt(), *args.Data, accessList, false)
+		// Apply the transaction
+		tracer := vm.NewAccessListTracer(accessList)
+		config := vm.Config{Tracer: tracer, Debug: true}
+		vmenv, _, err := b.GetEVM(ctx, msg, statedb, header, &config)
+		if err != nil {
+			return nil, 0, nil, err
+		}
+		res, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+		if err != nil {
+			return nil, 0, nil, fmt.Errorf("failed to apply transaction: %v err: %v", args.toTransaction().Hash(), err)
+		}
+		resList := tracer.RawAccessList()
+		if accessList.Equal(resList) {
```

Comment:
When you construct the types.AccessList from your custom internal type, the order is random, so `.Equal` will almost always compare apples to oranges.

https://github.com/ethereum/go-ethereum/pull/22550/files#diff-036bbd65d720bfc35dc772b0b472a1bbf9c23a47bf196c565ddf72e7decf0bc9R95 this is random iteration order

this one too https://github.com/ethereum/go-ethereum/pull/22550/files#diff-036bbd65d720bfc35dc772b0b472a1bbf9c23a47bf196c565ddf72e7decf0bc9R101

so https://github.com/ethereum/go-ethereum/pull/22550/files#diff-c426ecd2f7d247753b9ea8c1cc003f21fa412661c1f967d203d4edf8163da344R1368 even for the same access list will potentially compare different final sets